src/chapter-03/src/core.rs

```diff
@@ -25,28 +13,30 @@ pub struct ShortenUrlResponse {
     shortened_url: String,
 }
 
-pub struct UrlShortener {}
+#[derive(Default, Debug)]
+pub struct UrlShortener {
+    urls: RwLock<HashMap<String, String>>,
+}
 
 impl UrlShortener {
     pub fn new() -> Self {
-        Self {}
+        Self::default()
     }
 
     pub fn shorten_url(&self, req: ShortenUrlRequest) -> Result<ShortenUrlResponse, ()> {
         let short_url = self.generate_short_url();
 
-        let mut map = SHORTENED_URLS.lock().unwrap();
+        let mut map = self.urls.write().unwrap();
         map.insert(short_url.clone(), req.url_to_shorten);
 
         Ok(ShortenUrlResponse {
             shortened_url: short_url,
         })
     }
 
-    pub fn retrieve_url(&self, short_url: String) -> Option<String> {
-        let map = SHORTENED_URLS.lock().unwrap();
-        
-        map.get(&short_url).cloned()
+    pub fn retrieve_url(&self, short_url: &str) -> Option<String> {
```

Comment:
This is not majorly important but it's kind of a best practice AFAIK. Every time you need to receive a string in a function it's better to get a `&str` rather than a `String`. This is because then whoever is calling the function can avoid an allocation. In our case, this means we remove a lot of `.to_string()` on the calling side in `main`.

Similarly, if you ever need to receive a `Vec<T>` you should probably use `&[T]` in the function signature just to be a bit more flexible and generic and possibly allow the caller to avoid extra allocations.

Reply:
Of course, this is not a hard rule. In some cases you need to pass ownership to the function, so in these cases it makes sense to accept `String` or `Vec`